daemon/api_download.go

```diff
@@ -68,38 +108,154 @@ func postSnapDownload(c *Command, r *http.Request, user *auth.UserState) Respons
 }
 
 func streamOneSnap(c *Command, action snapDownloadAction, user *auth.UserState) Response {
-	actions := []*store.SnapAction{{
-		Action:       "download",
-		InstanceName: action.SnapName,
-		Revision:     action.Revision,
-		CohortKey:    action.CohortKey,
-		Channel:      action.Channel,
-	}}
-	sars, err := getStore(c).SnapAction(context.TODO(), nil, actions, user, nil)
+	secret, err := downloadTokensSecret(c)
 	if err != nil {
-		return errToResponse(err, []string{action.SnapName}, InternalError, "cannot download snap: %v")
+		return InternalError(err.Error())
+	}
+	theStore := getStore(c)
+
+	var ss *snapStream
+	if action.ResumeToken == "" {
+		var info *snap.Info
+		actions := []*store.SnapAction{{
+			Action:       "download",
+			InstanceName: action.SnapName,
+			Revision:     action.Revision,
+			CohortKey:    action.CohortKey,
+			Channel:      action.Channel,
+		}}
+		sars, err := theStore.SnapAction(context.TODO(), nil, actions, user, nil)
+		if err != nil {
+			return errToResponse(err, []string{action.SnapName}, InternalError, "cannot download snap: %v")
+		}
+		if len(sars) != 1 {
+			return InternalError("internal error: unexpected number %v of results for a single download", len(sars))
+		}
+		info = sars[0].Info
+
+		ss, err = newSnapStream(action.SnapName, info, secret)
+		if err != nil {
+			return InternalError(err.Error())
+		}
+	} else {
+		var err error
+		ss, err = newResumingSnapStream(action.SnapName, action.ResumeToken, secret)
+		if err != nil {
+			return BadRequest(err.Error())
+		}
+		ss.resume = action.resumePosition
 	}
-	if len(sars) != 1 {
-		return InternalError("internal error: unexpected number %v of results for a single download", len(sars))
+
+	if !action.HeaderPeek {
+		r, s, err := theStore.DownloadStream(context.TODO(), action.SnapName, ss.Info, action.resumePosition, user)
```

Comment:
I find the number of one-character variables hard to follow sometimes. Perhaps this could just be `stream, status, err := ...`?